File: patterns-use-cases/payment-state-machine/payment-state-machine-typescript/src/utils/accounts.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of Restate Examples for the Node.js/TypeScript SDK,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/examples/blob/main/LICENSE
 */

import * as restate from "@restatedev/restate-sdk";

//
// A simple virtual object, to track accounts.
// This is for simplicity to make this example work self-contained.
// This should be a database in a real scenario
//

type Result = { success: boolean; reason?: string };

export const userAccountObjects = restate.keyedRouter({
  deposit: async (ctx: restate.KeyedContext, accountId: string, amountCents: number) => {
    if (amountCents < 0) {
      throw new restate.TerminalError("Amount must not be negative");
    }

    const balanceCents: number = (await ctx.get<number>("balance")) ?? initializeRandomAmount();
    ctx.set("balance", balanceCents + amountCents);
  },

  withdraw: async (
    ctx: restate.KeyedContext,
    accountId: string,
    amountCents: number
  ): Promise<Result> => {
    if (amountCents < 0) {
      throw new restate.TerminalError("Amount must not be negative");
    }

    const balanceCents = (await ctx.get<number>("balance")) ?? initializeRandomAmount();

    if (balanceCents < amountCents) {
      return {
        success: false,
        reason: `balance to low: ${balanceCents} cents`,
      };
    }

    ctx.set("balance", balanceCents - amountCents);
    return { success: true };
  },
});

export type Api = typeof userAccountObjects;
export const api: restate.ServiceApi<Api> = { path: "accounts" };

// ----------------------------------------------------------------------------
//  miscellaneous utils
// ----------------------------------------------------------------------------

function initializeRandomAmount(): number {
  return Math.floor(Math.random() * 100_000 + 100_000);
}

File: python/patterns-use-cases/src/statemachinepayments/accounts/accounts.py
from restate import ObjectContext, VirtualObject
from restate.exceptions import TerminalError
import random

from src.statemachinepayments.types import Result

# A simple virtual object, to track accounts.
# This is for simplicity to make this example work self-contained.
# This should be a database in a real scenario
account = VirtualObject("account")

# The key under which we store the balance
BALANCE = "balance"


@account.handler()
async def deposit(ctx: ObjectContext, amount_cents: int):
    if amount_cents <= 0:
        raise TerminalError("Amount must be greater than 0")

    balance_cents = await ctx.get(BALANCE) or initialize_random_amount()
    ctx.set(BALANCE, balance_cents + amount_cents)


@account.handler()
async def withdraw(ctx: ObjectContext, amount_cents: int) -> Result:
    if amount_cents <= 0:
        raise TerminalError("Amount must be greater than 0")

    balance_cents = await ctx.get(BALANCE) or initialize_random_amount()
    if balance_cents < amount_cents:
        return Result(success=False, message=f"Insufficient funds: {balance_cents} cents")

    ctx.set(BALANCE, balance_cents - amount_cents)
    return Result(success=True, message="Withdrawal successful")


def initialize_random_amount() -> int:
    return random.randint(100_000, 200_000)
